default/AI/ResearchAI.py

```diff
@@ -1,30 +1,328 @@
+    """Get tech priority. 1 is default. 0 if not useful (but doesn't hurt to research), < 0 to prevent AI to research it
+    """
 
-def exclude_tech(tech_name):
-    return ((foAI.foAIstate.aggression < AIDependencies.TECH_EXCLUSION_MAP_1.get(tech_name, fo.aggression.invalid)) or
-            (foAI.foAIstate.aggression > AIDependencies.TECH_EXCLUSION_MAP_2.get(tech_name, fo.aggression.maniacal)) or
-            tech_name in TechsListsAI.unusable_techs())
-            
+    index = get_empire_hash()
+
+    immediate = 999 # i.e. research this tech NOW!
+    useless = 0 # useless for AI, mostly because the AI doesn't know how to use them
+    default = 1
+    unresearchable = -1 # anything unresearchable should use this TODO obtain this information from techs.txt
+
+    theory = 0 # theories has 0 priority as they don't unlock anything. If it ends up prereq for any other tech those tech will add the theories in.
+
+    growth = 2 # maximum population growth has higher priority.
+    production_boost = 1.5 # production boosts have higher priority. This include buildings that boosts production
+    research_boost = 2 # research boosts have higher priority. This include buildings that boosts research
+    meter_boost = 1
+    shield = 1.25 # shields are more powerful than armors or weapons
+    defensive = 2 if foAI.foAIstate.aggression <= fo.aggression.cautious else (1 if fo.currentTurn() > 10 else 0.5)
+    supply = 2 if ColonisationAI.galaxy_is_sparse() else 0.5
+
+    red_star = len(AIstate.empireStars.get(fo.starType.red, [])) != 0
+    bright_star = len(AIstate.empireStars.get(fo.starType.blue, [])) != 0
+    blackhole = len(AIstate.empireStars.get(fo.starType.blackHole, [])) != 0
+    neutron = len(AIstate.empireStars.get(fo.starType.neutron , [])) != 0
+
+    hull = 1
+    offtrack_hull = 0.05
+    offtrack_subhull = 0.25
+    # select one hull and specialize it, but some AI may want to have more hulls, by random
+    org = hull if index & 1 == 0 or (index + 97) % 227 < 21 else offtrack_hull
+    robotic = hull if index & 1 == 1 or (index + 107) % 229 < 22 else offtrack_hull
+    if ColonisationAI.got_ast:
+        extra = (index + 113) % 223 < 11
+        asteroid = hull if index & 2 == 0 or extra else offtrack_hull
+        if asteroid == hull and not extra:
+            org = offtrack_hull
+            robotic = offtrack_hull
+    else:
+        asteroid = offtrack_hull
+    if bright_star or blackhole:
+        extra = (index + 103) % 211 < 10
+        energy = hull if index & 4 == 0 or extra else offtrack_hull
+        if energy == hull and not extra:
+            org = offtrack_hull
+            robotic = offtrack_hull
+            asteroid = offtrack_hull
+    else:
+        energy = offtrack_hull
+
+    # further specialization for robotic hulls
+    contgrav = hull if robotic == hull and index & 8 == 0 else offtrack_subhull
+    nanorobo = hull if robotic == hull and index & 8 == 1 else offtrack_subhull
```

Comment:
Typying in Python:
```>>> 8 & 8 
8```

So ```index & 8 == 1``` will always evaluate to False. A simple ```and index & 8```seems to be your intention here (as 0 is interpreted as False)?